File: src/main/ruby/performance.rb
#!/usr/bin/ruby

class Move
	def initialize(from, jumped, to)
		print("Creating new move #{from} -> #{jumped} -> #{to}\n")
		@from = from
		@jumped = jumped
		@to = to
	end
	
	def to_s
		return "#{from} -> #{jumped} -> #{to}"
	end
	
	attr_reader :from, :jumped, :to
	
end

class Coordinate
	def initialize(row, hole)
        raise "Illegal hole number: #{hole} < 1" if (hole < 1)
        raise "Illegal hole number: #{hole} on row #{row}" if (hole > row)
		@row = row
		@hole = hole
	end
	
    def possibleMoves(rowCount)
	    moves = []
        
        # upward (needs at least 2 rows above)
        if (@row >= 3)
            
            # up-left
            if (@hole >= 3)
                moves.push(Move.new(
                        self,
                        Coordinate.new(@row - 1, @hole - 1),
                        Coordinate.new(@row - 2, @hole - 2)))
            end
            
            # up-right
            if (@row - @hole >= 2)
                moves.push(Move.new(
                        self,
                        Coordinate.new(@row - 1, @hole),
                        Coordinate.new(@row - 2, @hole)))
            end
        end
        
        # leftward (needs at least 2 pegs to the left)
        if (@hole >= 3)
            moves.push(Move.new(
                    self,
                    Coordinate.new(@row, @hole - 1),
                    Coordinate.new(@row, @hole - 2)))
        end
        
        # rightward (needs at least 2 holes to the right)
        if (@row - @hole >= 2)
            moves.push(Move.new(
                    self,
                    Coordinate.new(@row, @hole + 1),
                    Coordinate.new(@row, @hole + 2)))
        end

        # downward (needs at least 2 rows below)
        if (rowCount - @row >= 2)
            
            # down-left (always possible when there are at least 2 rows below)
            moves.push(Move.new(
                    self,
                    Coordinate.new(@row + 1, @hole),
                    Coordinate.new(@row + 2, @hole)))
            
            # down-right (always possible when there are at least 2 rows below)
            moves.push(Move.new(
                    self,
                    Coordinate.new(@row + 1, @hole + 1),
                    Coordinate.new(@row + 2, @hole + 2)))
        end
        
        return moves
    end

	def to_s()
		return "r#{@row}h#{@hole}"
	end
	
	def eql?(other)
		@row == other.row && @hole == other.hole
	end

	def ==(other)
		eql?(other)
	end
	
	attr_reader :row, :hole
	
end

class GameState

	def initialize(rows, emptyHole)
        @rowCount = rows
        @occupiedHoles = []
        for row in 1..rows
            for hole in 1..row
                def peg = Coordinate.new(row, hole)
                if !(peg == emptyHole)
                    occupiedHoles.push peg
                end
            end
        end
    end

    private GameState(GameState initialState, Move applyMe) {
        rowCount = initialState.rowCount;
        occupiedHoles = new ArrayList<Coordinate>(initialState.occupiedHoles);
        if (!occupiedHoles.remove(applyMe.getFrom())) {
            throw new IllegalArgumentException(
                    "Move is not consistent with game state: 'from' hole was unoccupied.");
        }
        if (!occupiedHoles.remove(applyMe.getJumped())) {
            throw new IllegalArgumentException(
                    "Move is not consistent with game state: jumped hole was unoccupied.");
        }
        if (occupiedHoles.contains(applyMe.getTo())) {
            throw new IllegalArgumentException(
                    "Move is not consistent with game state: 'to' hole was occupied.");
        }
        if (applyMe.getTo().getRow() > rowCount || applyMe.getTo().getRow() < 1) {
            throw new IllegalArgumentException(
                    "Move is not legal because the 'to' hole does not exist: " + applyMe.getTo());
        }
        occupiedHoles.add(applyMe.getTo());
    }
    
    public List<Move> legalMoves() {
        List<Move> legalMoves = new ArrayList<Move>();
        for (Coordinate c : occupiedHoles) {
            Collection<Move> possibleMoves = c.possibleMoves(rowCount);
            Iterator<Move> it = possibleMoves.iterator();
            while (it.hasNext()) {
                Move m = it.next();
                if (occupiedHoles.contains(m.getJumped()) && !occupiedHoles.contains(m.getTo())) {
                    legalMoves.add(m);
                }
            }
        }
        return legalMoves;
    }
    
    public GameState apply(Move move) {
        return new GameState(this, move);
    }

    public int pegsRemaining() {
        return occupiedHoles.size();
    }

    /**
     * Returns the full board state in a multiline string arranged to resemble a
     * real board. '*' characters signify occupied holes, and 'O' characters
     * signify empty ones.
     */
    @Override
    public String toString() {
        final String nl = System.getProperty("line.separator");
        StringBuilder sb = new StringBuilder();
        sb.append("Game with " + pegsRemaining() + " pegs:" + nl);
        for (int row = 1; row <= rowCount; row++) {
            int indent = rowCount - row;
            for (int i = 0; i < indent; i++) {
                sb.append(" ");
            }
            for (int hole = 1; hole <= row; hole++) {
                if (occupiedHoles.contains(new Coordinate(row, hole))) {
                    sb.append(" *");
                } else {
                    sb.append(" O");
                }
            }
            sb.append(nl);
        }
        return sb.toString();
    }
end

c1 = Coordinate.new(2,1)
c2 = Coordinate.new(2,1)

print "c1: #{c1}\n"
print "c1 moves: #{c1.possibleMoves(5)}\n"
print "c2: #{c2}\n"

print "c2 == c1: #{c2 == c1}\n"
print "c2.eql?(c1): #{c2.eql?(c1)}\n"

a = [c1, c2]
print a

File: src/main/python/performance.py
#!/usr/bin/env python2

# 
# performance.py
# 
# Copyright (c) 2010, Jonathan Fuerth
# 
# All rights reserved.
# 
# Redistribution and use in source and binary forms, with or without
# modification, are permitted provided that the following conditions
# are met:
# 
#     * Redistributions of source code must retain the above copyright
#       notice, this list of conditions and the following disclaimer.
#     * Redistributions in binary form must reproduce the above copyright
#       notice, this list of conditions and the following disclaimer in
#       the documentation and/or other materials provided with the
#       distribution.
#     * Neither the name of Jonathan Fuerth nor the names of other
#       contributors may be used to endorse or promote products derived
#       from this software without specific prior written permission.
# 
# THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
# "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
# LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
# A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
# OWNER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
# SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
# LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
# DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
# THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
# (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
# OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

from collections import namedtuple

class Move(namedtuple("Move", "fromh jumped to")):
    def __str__(self):
        return str(self.fromh) + " -> " + str(self.jumped) + " -> " + str(self.to)

class Coordinate(namedtuple("Coordinate", "row hole")):
    def legalMoves(self, rowCount, occupiedHoles, __moves=[(2, 0), (-2, 0), (2, 2), (-2, -2), (0, 2), (0, -2)]):
        row = self.row
        hole = self.hole

        for rOff, hOff in __moves:
            newRow = row + rOff
            newHole = hole + hOff
            if newRow < 1 or newHole < 1 or newRow < newHole or newRow > rowCount:
                continue
            t = (newRow, newHole)
            if t in occupiedHoles:
                continue
            j = (row + rOff // 2, hole + hOff // 2)
            if j not in occupiedHoles:
                continue
            yield Move(
                self,
                Coordinate(*j),
                Coordinate(*t),
            )

    def __str__(self):
        return "r" + str(self.row) + "h" + str(self.hole)


class GameState(object):

    def __init__(self, rows, emptyHole, initialState=None, applyMe=None):

        if initialState != None:
            # top-secret constructor overload for applying a move
            self.rowCount = initialState.rowCount
            self.occupiedHoles = set()
            self.occupiedHoles.update(initialState.occupiedHoles)

            # Note to those comparing this implementation to the others:
            # List.remove() raises ValueError if thr requested item is
            # not present, so the explicit errors are not raised here.
            # The self-checking nature of this method is still intact.

            self.occupiedHoles.remove(applyMe.fromh)
            self.occupiedHoles.remove(applyMe.jumped)

            if applyMe.to in self.occupiedHoles:
                raise RuntimeError, "Move is not consistent with game state: 'to' hole was occupied."

            if (applyMe.to.row > self.rowCount or applyMe.to.row < 1):
                raise RuntimeError, "Move is not legal because the 'to' hole does not exist: " + str(applyMe.to)

            self.occupiedHoles.add(applyMe.to)

        else:
            # normal constructor that sets up board
            self.rowCount = rows;
            self.occupiedHoles = set()
            for row in range(1, rows + 1):
                for hole in range(1, row + 1):
                    peg = Coordinate(row, hole)
                    if (not peg == emptyHole):
                        self.occupiedHoles.add(peg)

    def legalMoves(self):
        for c in self.occupiedHoles:
            for m in c.legalMoves(self.rowCount, self.occupiedHoles):
                yield m
    
    def applyMove(self, move):
        return GameState(None, None, self, move)

    def pegsRemaining(self):
        return len(self.occupiedHoles)

    def __str__(self):
        sb = []
        sb.append("Game with " + str(self.pegsRemaining()) + " pegs:\n")
        for row in range(1, self.rowCount + 1):
            indent = self.rowCount - row
            for i in range(0, indent):
                sb.append(" ")
            for hole in range(1, row + 1):
                if Coordinate(row, hole) in self.occupiedHoles:
                    sb.append(" *")
                else:
                    sb.append(" O")
            sb.append("\n")
        return "".join(sb)


gamesPlayed = 0
solutions = []
    
def search(gs, moveStack):
    global gamesPlayed
    global solutions
    
    if (gs.pegsRemaining() == 1):
        #print("Found a winning sequence. Final state:")
        #print(gs);

        solutionCopy = []
        solutionCopy.extend(moveStack)
        solutions.append(solutionCopy)
        
        gamesPlayed += 1
        return

    for m in gs.legalMoves():
        nextState = gs.applyMove(m)
        moveStack.append(m)
        search(nextState, moveStack)
        moveStack.pop()

    gamesPlayed += 1

from time import time

startTime = time()
gs = GameState(5, Coordinate(3, 2))
search(gs, [])
endTime = time()

print "Games played:    %6d" % (gamesPlayed)
print "Solutions found: %6d" % (len(solutions))
print "Time elapsed:    %6dms" % ((endTime - startTime) * 1000)
